src/group_theory/burnside.lean

```diff
@@ -0,0 +1,153 @@
+begin
+  use assume g', ⟨g * g'.1 * g⁻¹, stabilizer_orbit_iff.1 g'.2⟩,         -- to_fun
+  use assume g', ⟨g⁻¹ * g'.1 * g, stabilizer_orbit_inv_iff.1 g'.2⟩,     -- inv_fun
+  all_goals { assume g',
+              apply subtype.eq',
+              simp,
+              rw mul_assoc },
+    { rw inv_mul_cancel_right,
+      exact inv_mul_cancel_left _ _ },
+    { rw mul_inv_cancel_right,
+      exact mul_inv_cancel_left _ _ }
+end
+
+variables G α
+def orbits : Type v := quotient $ orbit_rel G α
+
+local attribute [instance] mul_action.orbit_rel
+local attribute [reducible] lattice.fixed_points
+
+/-- A trivial equivalence used in the proof of Burnside Lemma -/
+def burnside_aux₁ :
+  (Σ g : G, lattice.fixed_points ((•) g : α → α)) ≃
+    (Σ (o : orbits G α) (x : {x // ⟦x⟧ = o}), stabilizer G x.val) :=
+{ to_fun  := λ y, ⟨⟦y.snd.val⟧, ⟨⟨y.snd.val, rfl⟩, ⟨y.fst, y.snd.property⟩⟩⟩,
+  inv_fun := λ y, ⟨y.snd.snd.val, ⟨y.snd.fst.val, y.snd.snd.property⟩⟩,
+  left_inv := λ ⟨g, ⟨x, h⟩⟩, rfl,
+  right_inv := λ ⟨o, ⟨⟨x, hx⟩, ⟨g, h⟩⟩⟩, by subst o
+}
+
+noncomputable theory
+open classical
+
+variables {G α}
+
+/-- Choose an element of the group that sends `o.out` to a specific point of the orbit. -/
+def orbit_choice {o : orbits G α} (x : {x // ⟦x⟧ = o}) : G :=
+  some (quotient.exact $ x.2.trans o.out_eq.symm)
+
+def orbit_choice_spec {o : orbits G α} (x : {x // ⟦x⟧ = o}) :
+  orbit_choice x • o.out = x.val :=
+some_spec (quotient.exact $ x.2.trans o.out_eq.symm)
+
+/-- Disjoint union of stabilizers of all points of a given orbit
+    has the same cardinality as the group. -/
+def sigma_stabilizer_equiv_group (o : orbits G α) :
+  (Σ x : {x // ⟦x⟧ = o}, stabilizer G x.val) ≃ G :=
+begin
+  have h₁ : (Σ x : {x // ⟦x⟧ = o}, stabilizer G x.val) ≃
+    ({x // ⟦x⟧ = o} × (stabilizer G o.out)),
+    by { apply equiv.sigma_equiv_prod_of_equiv,
+         assume x,
+         rw [← (orbit_choice_spec x)],
+         symmetry,
+         apply equiv.stabilizer },
+
+  have h₂ : ({x // ⟦x⟧ = o} × (stabilizer G o.out)) ≃ ((orbit G o.out) × (stabilizer G o.out)),
+    by { apply equiv.cast,
+         congr,
+         funext,
+         apply propext,
+         rw [← quotient.out_eq o,quotient.eq,quotient.out_eq],
+         refl },
+
+  have h₃ : ((orbit G o.out) × (stabilizer G o.out)) ≃ G,
+    by { generalize: o.out = y,
+         symmetry,
+         apply (@mul_action.is_subgroup G _ _ _ y).group_equiv_quotient_times_subgroup.trans,
+         apply equiv.prod_congr,
+           { exact (orbit_equiv_quotient_stabilizer _ _).symm },
+           { refl } },
+
+  calc _ ≃ ({x // ⟦x⟧ = o} × (stabilizer G o.out))  : h₁
+    ...  ≃ ((orbit G o.out) × (stabilizer G o.out)) : h₂
+    ...  ≃ G                                        : h₃
+end
+
+/-- Equivalence behind Burnside Lemma.
+
+    Disjoint union of fixed points of all elements of the group is equivalent
+    to the product of the group and the orbits space. -/
+def burnside_equiv :
```

Comment:
mathlib has (to mathematicians) somewhat strange naming conventions. In particular, we don't call lemma's that have a "specific name" (e.g. named after a person) by that name. We give them the descriptive name that we also give all the helper lemmas. It is explicitly encourage to give the "maths name" in the comments, like you already do.
So this should probably be called `sigma_fixed_points_equiv_group_prod_orbits`.

Reply:
I think, it would be nice to have a file in `docs/` with a list of named theorems, and their mathlib names. Or an attribute (e.g., @[title=Burnside Lemma]) with some "search by title" tactic.
I renamed my lemmas/defs, and will push it soon.

Reply:
@urkud This is what the doc comments are for: cross referenced names, alternate names, and literature references (in addition to explanations of what the theorem is saying and relations to other formal theorems). Most theorems of course don't have all of these things, but that's the ideal.

The problem with calling a theorem "burnside's lemma" is that there may be many mathematically equivalent statements whose formal statements differ slightly, all of which would be called "burnside's lemma" by a mathematician shown the statement in isolation. The purpose of the formal name is to disambiguate these. If there is a reasonable argument that there is only one formal statement that could be given the name, and anyone thinking of the statement would call it "burnside" before a more direct name, then I would consider naming it something like `burnside`, but I don't think that applies in this case since there is another `burnside` right after this lemma, and it differs significantly in the formal statement.

Reply:
@digama0 Do I understand correctly that the best search engine for named theorems in mathlib is `git grep`? What do you think about adding it to `docs/`? There we can have them with wikipedia/scolarpedia links and/or textbook references.

P.S.: Renamed `burnside_equiv` and `burnside` according to `mathlib` naming conventions.

Reply:
@urkud Not quite. The `/-- -/` comments are machine readable, like Doxygen, although we currently aren't doing much with that information. But I don't see much of an advantage of another index in `docs` somewhere that can get out of date when there is already equivalent information interleaved with the definitions themselves, closer to the actual point of use.

Reply:
@digama0 As a potential user, I'd prefer to read some human-readable text about implementation of a specific theory in `mathlib` with links to specific facts than `/-- -/` comments. I mean, compare [sphinx](http://www.sphinx-doc.org/) based documentation to [doxygen](http://doxygen.nl/) or `coqdoc` output.
P.S.: I propose to close this branch of discussion as an off-topic in the context of the pull request. If I'll have some proposal, I'll come to Zulip, or open an RFC issue.

Reply:
(I'm not blocking this PR based on this discussion, although I haven't reviewed the rest of it in depth. I defer to the other maintainers.) There are several sphinx documentation projects on lean + mathlib, but they usually develop into standalone books, e.g. Theorem Proving in Lean or Logic & Proof. My hope is to have `/-- -/` comments serve as source material for a doxygen/coqdoc style output. Some people use coqdoc for literate programming style (a.k.a 50/50 full narrative documentation), which is probably a bit too verbose for mathlib except maybe on isolated definitions, but I know that people have played with these levels of documentation in tutorials and such; @kbuzzard @PatrickMassot @NeilStrickland have used this for teaching purposes and may have more pointers.

Reply:
@digama0 Thank you for explanation. BTW, would it be easier to review this PR, if I split it into a few based on what parts of the source tree I modify?

Reply:
PRs of <250 lines are generally a lot easier to review. I'll leave it up to you whether you actually want to split it. I think this is now actually quite close to being ready for merging. @ChrisHughes24 What do you think?